pkg/registration/spoke/managedcluster/claim_reconcile.go

```diff
@@ -46,31 +47,70 @@ func (r *claimReconcile) reconcile(ctx context.Context, cluster *clusterv1.Manag
 // exposeClaims saves cluster claims fetched on managed cluster into status of the
 // managed cluster on hub. Some of the customized claims might not be exposed once
 // the total number of the claims exceeds the value of `cluster-claims-max`.
-func (r *claimReconcile) exposeClaims(ctx context.Context, cluster *clusterv1.ManagedCluster) error {
+func (r *claimReconcile) exposeClaims(_ context.Context, cluster *clusterv1.ManagedCluster) error {
 	var reservedClaims, customClaims []clusterv1.ManagedClusterClaim
+	var clusterClaims []*clusterv1alpha1.ClusterClaim
 
 	// clusterClaim with label `open-cluster-management.io/spoke-only` will not be synced to managedCluster.Status at hub.
-	requirement, _ := labels.NewRequirement(labelCustomizedOnly, selection.DoesNotExist, []string{})
-	selector := labels.NewSelector().Add(*requirement)
-	clusterClaims, err := r.claimLister.List(selector)
+	requirement, err := labels.NewRequirement(labelCustomizedOnly, selection.DoesNotExist, []string{})
 	if err != nil {
-		return fmt.Errorf("unable to list cluster claims: %w", err)
+		return err
+	}
+	selector := labels.NewSelector().Add(*requirement)
+
+	if features.SpokeMutableFeatureGate.Enabled(ocmfeature.ClusterClaim) {
+		clusterClaims, err = r.claimLister.List(selector)
+		if err != nil {
+			return fmt.Errorf("unable to list cluster claims: %w", err)
+		}
 	}
 
 	// check if the cluster claim is one of the reserved claims or has a reserved suffix.
 	// if so, it will be treated as a reserved claim and will always be exposed.
 	reservedClaimNames := sets.New(clusterv1alpha1.ReservedClusterClaimNames[:]...)
 	reservedClaimSuffixes := sets.New(r.reservedClusterClaimSuffixes...)
+
+	// when ClusterProperties feature is not enabled, the informer will not be started and the lister will
+	// return an empty list, so we do not need to check featuregate here.
+	propertiesMap := map[string]*aboutv1alpha1.ClusterProperty{}
+	if features.SpokeMutableFeatureGate.Enabled(ocmfeature.ClusterProperty) {
+		clusterProperties, err := r.aboutLister.List(selector)
+		if err != nil {
+			return fmt.Errorf("unable to list cluster properties: %w", err)
+		}
+
+		for _, property := range clusterProperties {
+			propertiesMap[property.Name] = property
+		}
+	}
+
+	// convert claim to properties
```

Comment:
here is a little bit weird that we convert the claim to clusterProperty in line 93, and again we convert the clusterProperty to claim in line 104.

We can declare a `claimsMap := map[string]clusterv1.ManagedClusterClaim{}` at the beginning of the func, and for range the listed clusterclaims and clusterProperties, then covert both of them to clusterv1.ManagedClusterClaim and insert. finally we for range  the claimsMap directly to do the reserved claim matching.

Reply:
clusterv1.ManagedClusterClaim  is different from clusterv1alpha1.ClusterClaim. the former is the struct in managedcluster. The idea here is all clusterv1alpha1.ClusterClaim are converted to the properties, then we construct clusterv1.ManagedClusterClaim based on properties

Reply:
make sense, updated.